src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -262,6 +262,49 @@ FabricInfo * RetrieveCurrentFabric(CommandHandler * aCommandHandler)
     return Server::GetInstance().GetFabricTable().FindFabricWithIndex(index);
 }
 
+void FailSafeCleanup(FailSafeContext & failSafeContext)
+{
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Call to FailSafeCleanup");
+
+    // If an AddNOC or UpdateNOC command has been successfully invoked, terminate all CASE sessions associated with the Fabric
+    // whose Fabric Index is recorded in the Fail-Safe context (see ArmFailSafe Command) by clearing any associated Secure
+    // Session Context at the Server.
+    if (failSafeContext.NocCommandHasBeenInvoked())
+    {
+        Server::GetInstance().GetSecureSessionManager().ExpireAllPairingsForFabric(failSafeContext.GetFabricIndex());
```

Comment:
Because the event gets handled async, we could have had a new fail-safe context created since the event was posted, and then this fabric index is wrong.

Reply:
It will not get here in this case, I only proceed with the cleanup if 
"VerifyOrReturn(event->CommissioningComplete.PeerFabricIndex == failSafeContext.GetFabricIndex());"

Reply:
OK, but then we will miss doing cleanup that we should in fact do, no?

Reply:
That is right, since there is supposed to only be a single fail-safe period, and we share one failSafeContext, if it has been modified, then we can not do the proper clean-up. Any idea?

Reply:
To prevent failSafeContext get  used for anything when the event fires. we could set "mFailSafeArmed = false;" after CommissioningFailedTimerComplete event is processed by all subscribers, but this is not easy.

Reply:
Whatever state cleanup needs must be captured in the event, so that cleanup does not need to reference the failsafe context at all.